pkg/scheduler/framework/plugins/noderesources/fit.go

```diff
@@ -244,11 +247,121 @@ func (f *Fit) EventsToRegister() []framework.ClusterEventWithHint {
+	}
+
+	logger.V(5).Info("the max request resources of another scheduled pod got reduced and it may make the unscheduled pod schedulable", "pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+	return framework.Queue, nil
+}
+
+// isResourceScaleDown checks whether the resource request of the modified pod is less than the original pod
+// for the resources requested by the pod we are trying to schedule.
+func (f *Fit) isResourceScaleDown(targetPod, originalOtherPod, modifiedOtherPod *v1.Pod) bool {
+	if modifiedOtherPod.Spec.NodeName == "" {
+		// no resource is freed up whatever the pod is modified.
+		return false
+	}
+
+	// the other pod was scheduled, so modification or deletion may free up some resources.
+	originalMaxResourceReq, modifiedMaxResourceReq := &framework.Resource{}, &framework.Resource{}
+	originalMaxResourceReq.SetMaxResource(resource.PodRequests(originalOtherPod, resource.PodResourcesOptions{InPlacePodVerticalScalingEnabled: f.enableInPlacePodVerticalScaling}))
+	modifiedMaxResourceReq.SetMaxResource(resource.PodRequests(modifiedOtherPod, resource.PodResourcesOptions{InPlacePodVerticalScalingEnabled: f.enableInPlacePodVerticalScaling}))
+
+	// check whether the resource request of the modified pod is less than the original pod.
+	podRequests := resource.PodRequests(targetPod, resource.PodResourcesOptions{InPlacePodVerticalScalingEnabled: f.enableInPlacePodVerticalScaling})
+	for rName, rValue := range podRequests {
+		if rValue.IsZero() {
+			// We only care about the resources requested by the pod we are trying to schedule.
+			continue
+		}
+		switch rName {
+		case v1.ResourceCPU:
+			if originalMaxResourceReq.MilliCPU > modifiedMaxResourceReq.MilliCPU {
+				return true
+			}
+		case v1.ResourceMemory:
+			if originalMaxResourceReq.Memory > modifiedMaxResourceReq.Memory {
+				return true
+			}
+		case v1.ResourceEphemeralStorage:
+			if originalMaxResourceReq.EphemeralStorage > modifiedMaxResourceReq.EphemeralStorage {
+				return true
+			}
+		default:
+			if schedutil.IsScalarResourceName(rName) && originalMaxResourceReq.ScalarResources[rName] > modifiedMaxResourceReq.ScalarResources[rName] {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// isSchedulableAfterNodeChange is invoked whenever a node added or changed. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (f *Fit) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	_, modifiedNode, err := schedutil.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		return framework.Queue, err
+	}
+
+	var isFitFunc = func(pod *v1.Pod, node *v1.Node) bool {
+		if node == nil {
+			return false
+		}
+		nodeInfo := framework.NewNodeInfo()
+		nodeInfo.SetNode(node)
+		return len(Fits(pod, nodeInfo)) == 0
+	}
+
+	// TODO: also check if the original node meets the pod's resource requestments once preCheck is completely removed.
```

Comment:
why would the original node matter?

Reply:
My head is a little confused now and I can't figure it out here. @sanposhiho Can you help me?

Reply:
So, after the removal of preCheck, we'd like to check here to something like:
`if !isFitFunc(pod, originalNode) && isFitFunc(pod, modifiedNode)`
so that we return Queue only when Node is added or Node's `.status.Allocatable` is updated.

But, preCheck is currently causing trouble like:

1. Node is added. But, it's filtered out by preCheck (due to node's unready or whatever) and the scheduling queue doesn't receive NodeAdded. (noderesourcefit cannot receive NodeAdded.)
2. Node is updated and ready now. But, this event is NodeUpdated.

If we implement `if !isFitFunc(pod, originalNode) && isFitFunc(pod, modifiedNode)` while we still have preCheck, we may miss the event that could make Pods schedulable in the above scenario (due to (2)).